test/fgpu/test_send.py

```diff
@@ -0,0 +1,285 @@
+        )
+    return sensors
+
+
+def _fill_random(data: np.ndarray, rng: np.random.Generator) -> None:
+    """Fill an array with random bytes.
+
+    The underlying bytes are filled, rather than assuming any particular dtype.
+    """
+    view = data.view(np.uint8)
+    view[:] = rng.integers(0, 256, view.shape, dtype=np.uint8)
+
+
+def test_bad_substreams():
+    """Test that :class:`Chunk` raises an exception if n_substreams doesn't divide n_channels.
+
+    This is really just for code coverage; nothing really depends on it.
+    """
+    dtype = gaussian_dtype(8)
+    with pytest.raises(ValueError):
+        Chunk(
+            np.zeros((N_FRAMES, N_CHANNELS, N_SPECTRA_PER_HEAP, N_POLS), dtype),
+            np.zeros((N_FRAMES, N_POLS), np.uint32),
+            n_substreams=5,
+            feng_id=FENG_ID,
+            spectra_samples=SPECTRA_SAMPLES,
+        )
+
+
+async def test_send(
+    sample_bits: int,
+    queues: dict[str, list[spead2.InprocQueue]],
+    send_streams: list["spead2.send.asyncio.AsyncStream"],
+    recv_streams: dict[str, list[spead2.recv.asyncio.Stream]],
+    chunks: list[Chunk],
+    sensors: aiokatcp.SensorSet,
+    time_converter: TimeConverter,
+) -> None:
+    """Test sending data via the :class:`katgpucbf.fgpu.send.Chunk` interface.
+
+    .. todo::
+
+       This is a very limited test that does not cover sensors, metrics or
+       missing data.
+    """
+    # Send descriptors to all the streams
+    descriptor_heap = make_descriptor_heap(
+        channels_per_substream=N_CHANNELS // N_SUBSTREAMS,
+        spectra_per_heap=N_SPECTRA_PER_HEAP,
+        sample_bits=sample_bits,
+    )
+    for send_stream in send_streams:
+        for substream in range(N_SUBSTREAMS):
+            await send_stream.async_send_heap(descriptor_heap, substream_index=substream)
+
+    rng = np.random.default_rng(seed=1)
+    first_timestamp = 0x123456780000
+    skip_frames = 3
+    for i, chunk in enumerate(chunks):
+        _fill_random(chunk.data, rng)
+        # Note: this doesn't correspond in any way to the values in data.
+        # That isn't necessary for this test.
+        _fill_random(chunk.saturated, rng)
+        chunk.present[:] = True
+        timestamp = first_timestamp + i * SPECTRA_SAMPLES * N_SPECTRA_PER_HEAP * N_FRAMES
+        chunk.timestamp = timestamp
+        # Check that the property works as expected
+        assert chunk.timestamp == timestamp
+    # Knock out the first few frames, to test partial transmission
+    chunks[0].present[:skip_frames] = False
+    data = np.concatenate([chunk.data for chunk in chunks])
+    saturated = np.concatenate([chunk.saturated for chunk in chunks])
+    saturated = np.sum(saturated[skip_frames:], axis=0, dtype=np.uint64)
+
+    with PromDiff(namespace=METRIC_NAMESPACE) as prom_diff:
+        # Send all the chunks, without waiting for the first one to complete
+        # transmission (to ensure that the send streams have sufficient
+        # capacity).
```

Comment:
More for my own understanding, how does this ~shotgun-send ensure 
> that the send streams have sufficient capacity [?]

Would some underlying e.g. spead2 code complain if that weren't the case?

Reply:
Yes.